File: prototypes/AI_Deltas/LKB/LKB20250314A1B2/src/lens.js
import { config } from "./config.js";

export function lensMixer(metrics, weights = config.lensWeights) {
  const { energy = 0, dispersion = 0, biasAmplitude = 0 } = metrics;

  const predictivePush = weights.predictive * (0.4 + biasAmplitude * 0.6);
  const harmonicPull = weights.harmonic * dispersion;
  const systemicGuard = weights.systemic * Math.max(0, dispersion - 0.5);
  const humanBlend = weights.human * (0.2 + energy * 0.4);

  let pathB = config.basePathB + predictivePush - harmonicPull + humanBlend;
  pathB = clamp(pathB, config.harmonicClamp[0], config.harmonicClamp[1]);

  const damping = clamp(1 - (harmonicPull + systemicGuard) * config.forgivenessDamp, 0.4, 1);
  const biasScale = 1 + biasAmplitude * weights.predictive * 0.6;

  return { pathB, damping, biasScale };
}

export function clamp(value, min, max) {
  return Math.min(max, Math.max(min, value));
}

File: prototypes/AI_Deltas/LKB/260102193930_7LH8S4Z7/src/lens.py
"""Lens-inspired mixer translating metrics into gating signals."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Dict

from .config import LensWeights


@dataclass
class LensOutput:
    path_b_adjust: float
    damping: float
    bias_gain: float


def clamp(value: float, low: float, high: float) -> float:
    return max(low, min(high, value))


def mix_lenses(metrics: Dict[str, float], weights: LensWeights) -> LensOutput:
    """Blend lens weights into usable gating signals.

    metrics keys: energy, dispersion, bias, echo_gap.
    """
    energy = metrics.get("energy", 0.0)
    dispersion = metrics.get("dispersion", 0.0)
    bias_level = metrics.get("bias", 0.0)
    echo_gap = metrics.get("echo_gap", 0.0)

    human_term = max(0.0, 0.5 - dispersion)
    predictive_term = bias_level + echo_gap
    systemic_term = energy
    harmonic_term = dispersion

    path_b_adjust = clamp(
        (weights.predictive * predictive_term * 0.25)
        + (weights.systemic * systemic_term * 0.12)
        - (weights.human * human_term * 0.18),
        -0.25,
        0.25,
    )

    damping = clamp(harmonic_term * weights.harmonic * 0.7, 0.0, 0.6)

    bias_gain = clamp(1.0 + (weights.human * 0.2) - (weights.systemic * 0.08), 0.6, 1.4)

    return LensOutput(path_b_adjust=path_b_adjust, damping=damping, bias_gain=bias_gain)
